docs/micro-service-architecture.md

```diff
@@ -0,0 +1,248 @@
+- Null pointer exceptions can be a challenge.
+
+#### Kotlin:
+
+**Pros:**
+- Concise syntax reduces boilerplate code.
+- Modern language features, including null safety.
+- 100% interoperability with Java, easing migration.
+- Enhanced developer productivity.
+- Officially supported by JetBrains.
+
+**Cons:**
+- Smaller community compared to Java.
+- Learning curve for developers transitioning from Java.
+- Limited presence in legacy systems.
+
+### Github vs Gitlab
+
+#### GitHub:
+
+**Pros:**
+- Large community and ecosystem.
+- User-friendly interface.
+- Robust documentation.
+- Powerful collaboration features.
+- GitHub Actions for CI/CD.
+
+**Cons:**
+- Charges for private repositories.
+- Limited built-in CI/CD features.
+- Dependency on Microsoft.
+
+#### GitLab:
+
+**Pros:**
+- Integrated CI/CD.
+- Free private repositories.
+- Comprehensive DevOps platform.
+- Flexible deployment options.
+- Built-in container registry.
+
+**Cons:**
+- Learning curve for the interface.
+- Smaller community compared to GitHub.
+- Extensive feature set may overwhelm for simpler projects.
+
+### Keycloak vs Zitadel vs Auth0
+
+#### Keycloak
+**Pros:**
+- Open source
+- Widely use with a strong community
+- Wide range of features
+
+**Cons**
+- Complexity, when first trying keycloak it can be kind of complex to understand it's role and implications
+- Redhat documentation, keycloak documentation might be very hard to get through at some points due to it's documentation.
+
+#### Zitadel
+**Pros**
+- User-Centric Approach: Zitadel is designed with a strong focus on user experience and user-centric identity management.
+- Strong Security Features: Zitadel emphasizes security and includes features such as passwordless authentication and zero-trust architecture.
+- Enterprise-Grade: Zitadel positions itself as an enterprise-grade solution, catering to the needs of larger organizations.
+
+**Cons**
+- Lesser Adoption: As of my last knowledge update in January 2022, Zitadel might have had fewer users and 
+community support compared to more established solutions like Keycloak or Auth0.
+- Pricing: While Zitadel offers a free tier, pricing for larger enterprises might be a consideration.
+
+#### Auth0
+**Pros**
+- Ease of Use: Auth0 is known for its ease of use and quick setup. It provides a user-friendly dashboard and supports various identity scenarios out of the box.
+- Large User Base: Auth0 is widely adopted and has a large user base, leading to a robust community and extensive third-party integrations.
+- Extensive Documentation: Auth0 provides thorough and well-organized documentation, making it easier for developers to implement and troubleshoot.
+
+**Cons**
+- Pricing: Auth0's pricing might be higher compared to some other solutions, especially for large enterprises with high user counts.
+- Limited Customization in Free Tier: Some advanced features and customization options may be restricted in the free tier.
```

Comment:
I think you should merge all of that with their corresponding technologies above.